### PR TITLE
Implements bookings context and refactors manage bookings UI

### LINKED ISSUE #70: IN: Manage Bookings Refactor
Integrate the Bookings UI so that a sitter can accept/decline requests

### PR DESCRIPTION
**Pull request by michael-levitski**
### What this PR does:
- Closes #70 
- adds the ability for sitters to accept or decline bookings
- owners should be able to view the page as well but not modify the booking status
- Creates  `useBookingContext`
- The route for this feature is now a protected route

### Video:
https://www.loom.com/share/c498216a97094b0196701a6b294c4e8b

### Any information needed to test this feature (required):
- visit `/bookings` as a sitter
- create pending bookings

### Other info
- In the `Request` model I removed a duplicate `isSitter` field

### CODE REVIEW

**Comment by sundayezeilo** [comment]
Hey @michael-levitski, kindly link your ticket for this PR

File: server/models/Request.js [hunk 1]
```diff
@@ -19,6 +19,7 @@ const requestSchema = new Schema({
     required,
     validate: {
       validator: function (start) {
+        if (!this.isNew) return true;
```
Comment (ugur-akin) [inline server/models/Request.js]: I don't see a field called `isNew` here. Is this intentional? Can you elaborate if so?

File: server/models/Profile.js [hunk 2]
```diff
@@ -19,11 +19,6 @@ const profileSchema = new mongoose.Schema({
     required: true,
     default: false,
   },
-  isSitter: {
-    type: Boolean,
-    required: true,
-    default: false,
-  },
```
Comment (ugur-akin) [inline server/models/Profile.js:26]: Is this because we are now using mongoose discriminators? If not, can you explain why we are getting rid of this?

And just as a quick note, this is the sort of information that would be best placed in the PR description 👍🏼

File: server/controllers/request.js [hunk 3]
```diff
@@ -10,15 +11,25 @@ exports.getRequests = asyncHandler(async (req, res) => {
...
     "isSitter -_id"
   );
-  const selectedUserType = isSitter ? "sitter" : "owner";
-  const requests = await Request.find({
-    [selectedUserType]: req.user.id,
-  })
-    .populate("sitter")
-    .populate("owner")
+  const currentUserType = isSitter ? "sitter" : "owner";
+  const otherUserType = isSitter ? "owner" : "sitter";
```
Comment (ugur-akin) [inline server/controllers/request.js:15]: Now that we removed `isSitter` field, these won't work!

File: server/controllers/request.js [hunk 4]
```diff
@@ -10,15 +11,25 @@ exports.getRequests = asyncHandler(async (req, res) => {
...
     .sort("start")
     .exec();
 
+  const requests = queryResults.map((r) => {
+    const request = r.toJSON();
+    request.otherUser = request[otherUserType];
+    delete request[otherUserType];
+    return request;
+  });
+
```
Comment (ugur-akin) [inline server/controllers/request.js:32]: Is there a particular reason why we don't populate both `owner/sitter` and send the request that way? This seems a bit over-engineered at a first glance, would you kindly elaborate?

File: client/src/pages/Bookings/Bookings.tsx [hunk 5]
```diff
@@ -6,45 +6,25 @@ import PageContainer from '../../components/PageContainer/PageContainer';
...
+  if (loggedInUser === undefined || bookings === undefined) return <CircularProgress />;
   if (!loggedInUser) {
     history.push('/login');
     return <CircularProgress />;
   }
+
+  if (!bookings) {
+    updateSnackBarMessage('Error fetching bookings');
+    return <></>;
+  }
```
Comment (ugur-akin) [inline client/src/pages/Bookings/Bookings.tsx]: Optionally, I'd like to clean these checks up a little bit. `loggedInUser` logic should be handled by `ProtectedRoute`s (#28). We can focus on falling back on a spinner while we load bookings (e.g. maybe use `isLoading` in the context instead of comparing against `null/undefined`?)

File: client/src/interface/Booking.ts [hunk 6]
```diff
@@ -1,11 +1,11 @@
...
 export type Status = 'pending' | 'accepted' | 'declined' | 'paid';
+
 export interface Booking {
   _id: string;
   start: Date;
   end: Date;
   status: Status;
-  sitter: User;
-  owner: User;
+  otherUser: User;
```
Comment (ugur-akin) [inline client/src/interface/Booking.ts:10]: I'm personally against this change but I'd like to hear your reasoning as well 👍🏼

File: client/src/components/NextBookings/NextBooking.tsx [hunk 7]
```diff
@@ -1,29 +1,27 @@
...
         </Grid>
         <Typography sx={{ margin: '10px 0px' }}>
           {format(booking.start, 'd LLLL y, haaa')} - {format(booking.end, 'haaa')}
         </Typography>
         <Grid container direction="row" alignItems="center">
-          <Avatar src={`https://robohash.org/${booking.sitter.email}`}>{booking.sitter.name[0]}</Avatar>
+          <Avatar src={`https://robohash.org/${booking.otherUser.email}`}>{booking.otherUser.name[0]}</Avatar>
           <Typography sx={{ fontSize: '0.9rem', fontWeight: 'bold', marginLeft: '5px' }}>
-            {booking.sitter.name}
+            {booking.otherUser.name}
```
Comment (ugur-akin) [inline client/src/components/NextBookings/NextBooking.tsx:24]: Once again, I think it doesn't make too much sense to use the exact same page for both sitters and owners considering how different their stories/needs are. Would have been nice to see a page of its own designed for owners but that's my opinion 👍🏼

File: client/src/components/HandleRequestButton/HandleRequestButton.tsx [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+const HandleRequestButton = ({ bookingId }: PropTypes) => {
+  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);
+  const { setBookingStatus } = useBookings();
+  const { profile } = useAuth();
+  const { updateSnackBarMessage } = useSnackBar();
+
+  if (!profile.isSitter) {
+    return <></>;
+  }
```
Comment (ugur-akin) [inline client/src/components/HandleRequestButton/HandleRequestButton.tsx]: We seem to have removed this field from the model, just wanted to bring that to attention.

**Review by ugur-akin — COMMENTED** [review]
Hey @michael-levitski , good progress on this! I had some concerns about the implementation, so I asked for more clarification here and there. The context looks good overall though 🚀 

I want to reiterate how important clearly communicating your PRs are once again! As you will see from my questions, it can be very difficult to understand the reasoning behind some of the changes for someone who's not actively contributing to the same feature. Please be sure to format your PRs in the future so that:
- The title is descriptive and is in plain English (e.g. `Implements bookings context and refactors manage bookings UI`)
- Make sure to [link the issue with a keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) in your PR body, this will link it both in Github and Zenhub. If there is no issue for your task, you can create one with a brief description to help others understand what the PR is about. 

I'd also recommend going over [this article](https://hugooodias.medium.com/the-anatomy-of-a-perfect-pull-request-567382bb6067) for more tips about how to communicate your PRs effectively 👍🏼 

Cheers, 🎉

File: server/models/Request.js (hunk 1 above)
Comment (michael-levitski) [reply server/models/Request.js]: `isNew` is an inherited property for all mongoose schemas. It will be true on create, but false on update. Upon further thought, it looks like this line should be omitted

File: server/models/Profile.js (hunk 2 above)
Comment (michael-levitski) [reply server/models/Profile.js:26]: For some reason lines 22-26 were duplicate of lines 17-21 so I was fixing that as well. Thanks for the helpful PR feedback

File: server/controllers/request.js (hunk 4 above)
Comment (michael-levitski) [reply server/controllers/request.js:32]: based on team conversations, we wanted the same page to be used for owners and sitters

**Comment by sundayezeilo** [comment]
Hey @michael-levitski, please resolve merge conflicts in this PR

File: server/models/Profile.js (hunk 2 above)
Comment (ugur-akin) [reply server/models/Profile.js:26]: Yeah I missed this, thanks 👍🏼

File: server/models/Request.js (hunk 1 above)
Comment (ugur-akin) [reply server/models/Request.js]: Ah gotcha!

**Review by ugur-akin — COMMENTED** [review]
LGTM 👍🏼 Please resolve merge conflicts so we can go ahead and approve this

File: client/src/context/useBookingsContext.tsx [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    }
+  }, []);
+
+  useEffect(() => {
+    fetchBookings();
+  }, [fetchBookings]);
+
+  const setBookingStatus = useCallback(async (id: string, status: Status) => {
+    const booking = await updateBooking({ status }, id);
+    setBookings((prevBookings) => {
```
Comment (rajivtitus) [inline client/src/context/useBookingsContext.tsx]: We would want to update the bookings in our local state only if the `updateBooking` API call is successful

File: client/src/context/useBookingsContext.tsx [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+      setIsLoadingBookings(false);
+    }
+  }, []);
+
+  useEffect(() => {
+    fetchBookings();
+  }, [fetchBookings]);
+
+  const setBookingStatus = useCallback(async (id: string, status: Status) => {
+    const booking = await updateBooking({ status }, id);
```
Comment (rajivtitus) [inline client/src/context/useBookingsContext.tsx]: Let's also perform some basic error handling here and update the Snackbar accordingly

**Review by rajivtitus — APPROVED** [review]
LGTM! Thanks for making changes, just a couple of comments to address before merging 👍🏽
